### PR TITLE
[EIP-1559] Step 9 - Update transaction processing with EIP-1559 rules.

### PR DESCRIPTION
**Pull request by abdelhamidbakhta**
## PR description
- Created `TransactionPriceCalculator` to compute the transaction gas price.
- Created `CoinbaseFeePriceCalculator` to compute the coinbase reward.
- Updated `MainnetTransactionProcessor` to apply new consensus rules:
    - Updated transaction gas price for EIP-1559 transactions.
    - Updated coinbase reward for EIP-1559 transactions.
- Updated `MainnetProtocolSpecs` to build the correct TransactionProcessor for EIP-1559 definition.

## Fixed Issue(s)
- #601 
- #605 

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -160,59 +160,15 @@ Workaround -> When using onchain permissioning, ensure bootnodes are also valida
 ### Additions and Improvements
```
Comment (shemnon) [inline CHANGELOG.md:160]: This file may need a hand merge.  I don't think there should be any diffs.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/fees/CoinbaseFeePriceCalculator.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ */
+package org.hyperledger.besu.ethereum.core.fees;
+
+import org.hyperledger.besu.config.experimental.ExperimentalEIPs;
+import org.hyperledger.besu.ethereum.core.Gas;
+import org.hyperledger.besu.ethereum.core.Wei;
+
+import java.util.Optional;
+
+@FunctionalInterface
```
Comment (shemnon) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/fees/CoinbaseFeePriceCalculator.java:23]: 😄 This should produce compiler errors now if another method is added to the interface.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/fees/TransactionPriceCalculator.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  Wei price(Transaction transaction, Optional<Long> baseFee);
+
+  static TransactionPriceCalculator frontier() {
+    return (transaction, baseFee) -> transaction.getGasPrice();
+  }
+
+  static TransactionPriceCalculator eip1559() {
+    return (transaction, maybeBaseFee) -> {
+      ExperimentalEIPs.eip1559MustBeEnabled();
+      final long gasPremium = transaction.getGasPremium().orElseThrow().getValue().longValue();
```
Comment (shemnon) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/fees/TransactionPriceCalculator.java:34]: Will gasPremium and Fee Cap always be in legacy TXes in a mixed 1559/legacy state?  i.e. is there a fixed calculation for them based on gas price?  i.e. `orElseThrow` will only throw if there are config issues?

Comment (abdelhamidbakhta) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/fees/TransactionPriceCalculator.java:34]: orElseThrow will only throw if there are config issues. You cannot derived gasPremium and fee cap from gas price.

File: CHANGELOG.md (hunk 1 above)
Comment (abdelhamidbakhta) [reply CHANGELOG.md]: Hand merged.

**Pull request merged**